### ISSUE TITLE
HHVM support

### ISSUE THREAD

**Issue by thekid** [xp-framework/core#47]
After xp-runners is now basically [looking good](https://travis-ci.org/xp-framework/xp-runners/jobs/46112931), this issue gathers all tasks necessary to support HHVM.

## HHVM incompatibilities & bugs
Various [bugs](https://github.com/xp-framework/core/issues/47#issuecomment-68955937) in HHVM need to be addressed, workarounds found and PHP compatibility retained at the same time. Travis CI will be a great help here:)

## Hack language
HHVM brings [Hack language](http://hacklang.org/). It features stuff PHP will or might see at some day in the future, e.g. return types, annotations or function types; and others it will probably not, like generics, type aliases, XHP, to name a few. An overview of Hack language features can be found [here](http://docs.hhvm.com/manual/en/hack.intro.php)

*See also #36 - PHP 7 support*

**Comment by thekid** [comment]
Changed test for `core` to `standard` and remove `static` modifier to stream wrapper methods stream_metadata() and url_stat(), two down.

See https://travis-ci.org/xp-framework/core/jobs/46140841

**Comment by thekid** [comment]
> Runtime::getInstance()->newInstance() yields unsupported parameters

Now taken care of by prepending `--php` to the hhvm invocation to force parsing the rest of the command line as PHP-CLI would.

**Comment by thekid** [comment]
> __call and __callStatic have differing behavior - see src/test/config/unittest/core.ini 

Fixed by offsetting 1 when `defined('HHVM_VERSION')`. This is likely an internal implementation detail caused by a more optimized method lookup algorithm in HHVM.

**Comment by thekid** [comment]
> Caught Exception lang.XPException (DateTimeZone::__construct(): Unknown or bad timezone (UNKNOWN)) instead of expected lang.IllegalArgumentException 

Both `timezone_open()` and `date_create()` behave differently in HHVM than in PHP. There's an easy way around: Use DateTimeZone and DateTime's constructors... they don't have this problem.

**Comment by thekid** [comment]
> stream_metadata() in stream wrappers is invoked differently for chmod() 

This was actually a different problem - the undefined variable `$value` was raised because setting the `$context` member of the stream wrapper defined inside the net.xp_framework.unittest.logging.FileAppenderTest class failed. Fix this by declaring a member (like `io.streams.Streams` does.

**Comment by thekid** [comment]
> Stream filters don't work correctly, e.g. base64, gz, bz2 and so on

Handled momentarily by checking availability of filters in unittests, and skipping if not.

**Comment by thekid** [comment]
> Segfaults in net.xp_framework.unittest.io.DomApiStreamsTest

Fixed for the moment by ignoring them.

***:red_circle: needs work***

**Comment by thekid** [comment]
> In certain situations, typehints resolve differently - Typehint for net.xp_framework.unittest.reflection.MethodParametersTest::fixture()'s parameter "b" cannot be resolved: Class self does not exist 

This error doesn't occur in HHVM-nightly (3.6.0)

**Comment by thekid** [comment]
Merged #56 for the moment, we need to discuss whether base64 streams should work with HHVM.

**Comment by thekid** [comment]
```sh
vagrant@vagrant-ubuntu-utopic-64:/devel/xp/core$ cat Test.class.php
<?hh

use util\cmd\Console;

class Test extends \lang\Object {

  public static function main(): int {
    Console::writeLine('Hello World');
    return 0;
  }
}
vagrant@vagrant-ubuntu-utopic-64:/devel/xp/core$ xp Test
Hello World
```
